### PR TITLE
US 23-24 Shopping cart

### PR DESCRIPTION
**Pull request by helloeduardo**
- [x] done

User Story 23, Adding Item Quantity to Cart

As a visitor
When I have items in my cart
And I visit my cart
Next to each item in my cart
I see a button or link to increment the count of items I want to purchase
I cannot increment the count beyond the item's inventory size

- [x] done

User Story 24, Decreasing Item Quantity from Cart

As a visitor
When I have items in my cart
And I visit my cart
Next to each item in my cart
I see a button or link to decrement the count of items I want to purchase
If I decrement the count to 0 the item is immediately removed from my cart

### CODE REVIEW

File: app/controllers/cart_controller.rb [hunk 1]
```diff
@@ -23,5 +23,20 @@ def remove_item
     redirect_to '/cart'
   end
 
+  def increment
+    item = Item.find(params[:item_id])
+    if cart.items[item] < item.inventory
+      cart.add_item(item.id.to_s)
```
Comment (BJSherman80) [inline app/controllers/cart_controller.rb:29]: I like how you used the pre-existing add_item method for this

File: app/models/cart.rb [hunk 2]
```diff
@@ -32,4 +32,8 @@ def total
     end
   end
 
+  def decrement_item(item)
+    @contents[item.id.to_s] -= 1
```
Comment (BJSherman80) [inline app/models/cart.rb:36]: Great job putting this data logic into the model

File: app/models/item.rb [hunk 3]
```diff
@@ -25,4 +40,8 @@ def no_orders?
     item_orders.empty?
   end
 
+  def quantity_sold
+    item_orders.sum(:quantity)
+  end
+
```
Comment (BJSherman80) [inline app/models/item.rb:46]: straight forward and simple!

File: app/views/cart/show.html.erb [hunk 4]
```diff
@@ -30,7 +30,11 @@
       </td>
       <td>
         <center>
-          <p><%=quantity%></p>
+          <section class="quantity">
+            <%= button_to "+", "/cart/#{item.id}/increment", method: :patch %>
+            <p><%=quantity%></p>
```
Comment (BJSherman80) [inline app/views/cart/show.html.erb:35]: I like how this looks in development w/ the + and - above and below the quantity

File: app/views/items/index.html.erb [hunk 5]
```diff
@@ -2,24 +2,41 @@
...
+<section class="item-statistics">
+  <p>Item Statistics</p>
+  <p>Top Five Bought</p>
+  <ol class="top-five">
+    <% @items.top_five.each do |item| %>
+      <li><%= item.name %>: Quantity Bought: <%= item.quantity_sold %> </li>
+    <% end %>
+  </ol>
+  <p>Bottom Five Bought</p>
+  <ol class="bottom-five">
```
Comment (BJSherman80) [inline app/views/items/index.html.erb:14]: perfect positioning for the ol

File: spec/features/items/index_spec.rb [hunk 6]
```diff
@@ -47,15 +46,63 @@
...
+
+        within ".bottom-five" do
+          expect(page.all('li')[0].text).to eq("Gatorskins: Quantity Bought: 2")
+          expect(page.all('li')[1].text).to eq("Pull Toy: Quantity Bought: 5")
+          expect(page.all('li')[2].text).to eq("Chain: Quantity Bought: 6")
+          expect(page.all('li')[3].text).to eq("Dog Bone: Quantity Bought: 7")
+          expect(page.all('li')[4].text).to eq("Lizard Head: Quantity Bought: 8")
+        end
+
       end
```
Comment (BJSherman80) [inline spec/features/items/index_spec.rb:104]: Massive test but it looks great, good job at testing the sad path

**Review by BJSherman80 — APPROVED** [review]

**Review by the-color-bliu — APPROVED** [review]
Everything looks great! Really good job with ActiveRecord.

File: app/controllers/cart_controller.rb [hunk 7]
```diff
@@ -23,5 +23,20 @@ def remove_item
...
+      flash[:error] = "Cannot add item"
+    end
+    redirect_to '/cart'
+  end
+
+  def decrement
+    item = Item.find(params[:item_id])
+    cart.decrement_item(item)
+    redirect_to '/cart'
+  end
```
Comment (nmking22) [inline app/controllers/cart_controller.rb:40]: Let's add an if statement that guards against item number going into negative values.

**Review by nmking22 — APPROVED** [review]
Looks really clean. Just need to add that guard statement for decrement stopping at 0

**Pull request merged**